SteamKit2/SteamKit2/Steam/Handlers/SteamMatchmaking/Lobby.cs

```diff
@@ -0,0 +1,383 @@
+            /// </summary>
+            public SteamID OwnerSteamID { get; }
+
+            /// <summary>
+            /// The metadata of the lobby; string key-value pairs.
+            /// </summary>
+            public IReadOnlyDictionary<string, string> Metadata => metadata;
+
+            /// <summary>
+            /// The maximum number of members that can occupy the lobby.
+            /// </summary>
+            public int MaxMembers { get; }
+
+            /// <summary>
+            /// The number of member that are currently occupying the lobby.
+            /// </summary>
+            public int NumMembers { get; }
+
+            /// <summary>
+            /// A list of lobby members, this does not include the lobby owner.
+            /// </summary>
+            public IReadOnlyList<Member> Members => members;
+
+            /// <summary>
+            /// The distance of the lobby.
+            /// </summary>
+            public float? Distance { get; }
+
+            /// <summary>
+            /// The weight of the lobby.
+            /// </summary>
+            public long? Weight { get; }
+
+            internal List<Member> members;
+            internal Dictionary<string, string> metadata;
+
+            internal Lobby( SteamID steamId, ELobbyType lobbyType, int lobbyFlags, SteamID ownerSteamId, Dictionary<string, string> metadata, int maxMembers,
+                int numMembers, List<Member> members, float? distance, long? weight )
+            {
+                if ( members != null && members.Count != numMembers - 1 )
+                {
+                    throw new ArgumentException( "when members is non-null, members.Count must be equal to numMembers - 1" );
+                }
+
+                SteamID = steamId;
+                LobbyType = lobbyType;
+                LobbyFlags = lobbyFlags;
+                OwnerSteamID = ownerSteamId;
+                metadata = metadata ?? new Dictionary<string, string>();
+                MaxMembers = maxMembers;
+                NumMembers = numMembers;
+                members = members ?? new List<Member>();
+                Distance = distance;
+                Weight = weight;
+            }
+
+            /// <summary>
+            /// Return a new <see cref="Lobby"/> instance with equivalent data as this lobby, but is safe to
+            /// access on another thread.
+            /// </summary>
+            /// <returns>A new <see cref="Lobby"/> instance that is equivalent to this one.</returns>
+            public Lobby Clone()
+            {
+                return new Lobby(
+                    SteamID,
+                    LobbyType,
+                    LobbyFlags,
+                    OwnerSteamID,
+                    new Dictionary<string, string>( metadata ),
+                    MaxMembers,
+                    NumMembers,
+                    new List<Member>( members ), // We're not doing a deep copy as Member is never mutated.
```

Comment:
I don't see where we mutate either the dictionary or the list, so why create new ones?

Reply:
We mutate the list on line 342, granted I've refactored locally. It's purely immutable now.